core/src/DecoderResult.h

```diff
@@ -46,44 +48,54 @@ class DecoderResult
 	int _structuredAppendParity = 0;
 	std::shared_ptr<CustomData> _extra;
 
-public:
-	//explicit DecoderResult(DecodeStatus status);
-	//DecoderResult(const ByteArray& rawBytes, const String& text, std::list<ByteArray>& byteSegments, const std::string& ecLevel, int saSequence, int saParity);
-	//DecoderResult(const ByteArray& rawBytes, const String& text, std::list<ByteArray>& byteSegments, const std::string& ecLevel);
-	DecoderResult() {}
 	DecoderResult(const DecoderResult &) = delete;
 	DecoderResult& operator=(const DecoderResult &) = delete;
 
-	const ByteArray& rawBytes() const { return _rawBytes; }
-	void setRawBytes(const ByteArray& bytes) { _rawBytes = bytes; _numBits = 8 * bytes.length(); }
-	int numBits() const { return _numBits; }
-	void setNumBits(int numBits) { _numBits = numBits; }
-
-	const std::wstring& text() const { return _text; }
-	void setText(const std::wstring& txt) { _text = txt; }
-
-	const std::list<ByteArray>& byteSegments() const { return _byteSegments; }
-	void setByteSegments(const std::list<ByteArray>& segments) { _byteSegments = segments; }
-
-	std::wstring ecLevel() const { return _ecLevel; }
-	void setEcLevel(const std::wstring& level) { _ecLevel = level; }
-
-	int errorsCorrected() const { return _errorsCorrected; }
-	void setErrorsCorrected(int ec) { _errorsCorrected = ec; }
-
-	int erasures() const { return _erasures; }
-	void setErasures(int e) { _erasures = e; }
+public:
+	DecoderResult(DecodeStatus status) : _status(status) {}
+	DecoderResult(ByteArray&& rawBytes, std::wstring&& text) : _rawBytes(std::move(rawBytes)), _text(std::move(text)) {
+		_numBits = 8 * rawBytes.length();
+	}
+
+	DecoderResult() = default;
+	DecoderResult(DecoderResult&&) = default;
+	DecoderResult& operator=(DecoderResult&&) = default;
+
+	bool isValid() const { return StatusIsOK(_status); }
+	DecodeStatus errorCode() const { return _status; }
+
+	const ByteArray& rawBytes() const & { return _rawBytes; }
+	ByteArray&& rawBytes() && { return std::move(_rawBytes); }
+	const std::wstring& text() const & { return _text; }
+	std::wstring&& text() && { return std::move(_text); }
+
+	// Simple macro to set up getter/setter methods that save lots of boilerplate.
+	// It sets up a standard 'const & () const', 2 setters for setting lvalues via
+	// copy and 2 for setting rvalues via move. They are provided each to work
+	// either on lvalues (normal 'void (...)') or on rvalues (returning '*this' as
+	// rvalue). The latter can be used to optionally initialize a temporary in a
+	// return statement, e.g.
+	//    return DecoderResult(bytes, text).setEcLevel(level);
+#define PROPERTY(TYPE, GETTER, SETTER) \
```

Comment:
OK, but we can name the macro ZX_PROPERTY instead which will greatly reduce the chance of name clashing.